### ISSUE TITLE
Remove poll_read / poll_write from TCP / UDP types

### ISSUE THREAD

**Issue by carllerche** [tokio-rs/tokio#75]
These types are legacy from the original `Io` trait implementation. There isn't really much of a point to having them and the exact guarantees that these functions provide are undefined (specifically, do they permit false positives or not?).

**Comment by alexcrichton** [comment]
I'd personally see these as important functions to have. IIRC we removed them from traits but specifically decided to leave them on the concrete types that can provide the implementation.

This is exposing in some sense capabilities about epoll where you can discover readability/writability of a socket without actually reading and/or writing. Examples also [use these functions](https://github.com/tokio-rs/tokio-socks5/blob/a6952adde41289267215c9cdd0487df025214952/src/main.rs#L564-L568). 

Unless they're problematic I would not remove them.

**Comment by carllerche** [comment]
@alexcrichton IMO the biggest issue is the semantics are not defined. When do they fire? What does `poll_read` returning ready mean? Does the socket have data? Is it closed? Did some unrelated event get rolled up with the read semantic? What are the guarantees? Are false negatives permitted (obviously not, but this isn't documented), will a `poll_read` guarantee that the next call to `read` will have data (this is how `poll_ready` on mpsc sender works)?

I personally have never needed these functions and I am not aware of a real use case that needs them (if anyone is aware of a counter example, please let me know).

In the examples you showed, I'm not sure what the benefit is of using poll_read / poll_write vs. just attempting the opts.

On top of that, I'm not actually sure how the example is actually correct. I see two bugs:

* [This line](https://github.com/tokio-rs/tokio-socks5/blob/a6952adde41289267215c9cdd0487df025214952/src/main.rs#L598) does `try_nb` which will skip the write if the socket is !readable but is writable.
* [This line](https://github.com/tokio-rs/tokio-socks5/blob/a6952adde41289267215c9cdd0487df025214952/src/main.rs#L609) does not call `try_nb` which will result in a "not ready" scenario being treated as an error.

tl;dr:

* The guarantees of `poll_read` / `poll_write` are not defined.
* I have personally never needed the functions.
* I am not sure of a real use case that needs them.
* The example doesn't need those functions. The block of code that uses them can just be deleted to same effect.
* The example is buggy.

**Comment by alexcrichton** [comment]
@carllerche I think those are some good questions, yes, but I don't think it's necessary to have absolutely rock-solid answers to keep functions like this? For example I think they've all got pretty reasonable answers with the current behavior today and I'd imagine they'd need tweaking but we could do that over time as necessary.

In terms of use cases that's sort of what the example was trying to show. The example shows where you can wait for the precise moment a set of sockets reach a set of readiness conditions to actually do a more expensive operation like buffer allocation or something like that.

The example I linked [has some documentation](https://github.com/tokio-rs/tokio-socks5/blob/a6952adde41289267215c9cdd0487df025214952/src/main.rs#L570-L596) about how `poll_read` and `poll_write` are just optimizations, and can return readiness spuriously or otherwise not guarantee any particular readiness condition. As such it's documented that the bugs you pointed out are indeed bugs, but they're intended to be bugs that could be fixed at one point in the future. I didn't fix them at the time as I felt it would detract too much from what the example was trying to show.

I'm particularly quite worried about losing the functionality here, the example was just meant to be one instance of how these could be used. Thinking more though I believe the use cases go beyond just what I linked above. These functions also serve as proxies to the underlying readiness of the socket file descriptor. Crates like `tokio-uds` leverage this by [enabling external implementations](https://github.com/tokio-rs/tokio-uds/pull/22) of things like `sendmsg` and `recvmsg` on Unix sockets. Now I realize that this is for UDP/TCP types of course and not Unix sockets, but I think that something like `poll_{read,write}` is required to bind functionality otherwise not exposed in Tokio itself.

For example I think if you wanted to wait until [`peek` was ready to go](https://github.com/carllerche/mio/commit/ed3d66af5b71989e28e322dd9abbc7b29ae4b9aa) you'd have to use these functions right now. We can of course add bindings over time for functions like this, but I'd worry that each OS would have its own flavorful form of functionality which we may not always want to expose.

**Comment by carllerche** [comment]
This happend via another PR.

I believe that the functionality will come back in the future, but I don't want to commit on the exact functions now especially given that `poll_read` / `poll_write` may be repurposed as names for other functionality.
